File: crates/openduo-tools/src/pipelines.rs
use crate::registry::Tool;
use anyhow::Result;
use async_trait::async_trait;
use openduo_core::gitlab_client::GitLabClient;
use serde_json::{json, Value};

pub struct PipelineTools;

impl PipelineTools {
    pub fn all(client: GitLabClient) -> Vec<Box<dyn Tool>> {
        vec![
            Box::new(ListPipelines {
                client: client.clone(),
            }),
            Box::new(GetPipeline {
                client: client.clone(),
            }),
            Box::new(TriggerPipeline {
                client: client.clone(),
            }),
            Box::new(RetryPipeline {
                client: client.clone(),
            }),
            Box::new(CancelPipeline {
                client: client.clone(),
            }),
            Box::new(GetPipelineJobs {
                client: client.clone(),
            }),
            Box::new(GetJobLog {
                client: client.clone(),
            }),
        ]
    }
}

struct ListPipelines {
    client: GitLabClient,
}
#[async_trait]
impl Tool for ListPipelines {
    fn name(&self) -> &str {
        "list_pipelines"
    }
    fn description(&self) -> &str {
        "List pipelines for a project."
    }
    fn parameters_schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "project_id": { "type": "string" },
                "per_page": { "type": "integer", "default": 20 }
            },
            "required": ["project_id"]
        })
    }
    async fn execute(&self, args: Value) -> Result<String> {
        let pid = urlencoding::encode(
            args["project_id"]
                .as_str()
                .ok_or_else(|| anyhow::anyhow!("project_id required"))?,
        );
        let per_page = args["per_page"].as_u64().unwrap_or(20);
        let v: Vec<Value> = self
            .client
            .get(&format!("projects/{}/pipelines?per_page={}", pid, per_page))
            .await?;
        Ok(serde_json::to_string_pretty(&v)?)
    }
}

struct GetPipeline {
    client: GitLabClient,
}
#[async_trait]
impl Tool for GetPipeline {
    fn name(&self) -> &str {
        "get_pipeline"
    }
    fn description(&self) -> &str {
        "Get details of a specific pipeline."
    }
    fn parameters_schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "project_id": { "type": "string" },
                "pipeline_id": { "type": "integer" }
            },
            "required": ["project_id", "pipeline_id"]
        })
    }
    async fn execute(&self, args: Value) -> Result<String> {
        let pid = urlencoding::encode(
            args["project_id"]
                .as_str()
                .ok_or_else(|| anyhow::anyhow!("project_id required"))?,
        );
        let pipeline_id = args["pipeline_id"]
            .as_u64()
            .ok_or_else(|| anyhow::anyhow!("pipeline_id required"))?;
        let v: Value = self
            .client
            .get(&format!("projects/{}/pipelines/{}", pid, pipeline_id))
            .await?;
        Ok(serde_json::to_string_pretty(&v)?)
    }
}

struct TriggerPipeline {
    client: GitLabClient,
}
#[async_trait]
impl Tool for TriggerPipeline {
    fn name(&self) -> &str {
        "trigger_pipeline"
    }
    fn description(&self) -> &str {
        "Trigger a new pipeline for a ref."
    }
    fn parameters_schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "project_id": { "type": "string" },
                "ref": { "type": "string", "description": "Branch or tag name" }
            },
            "required": ["project_id", "ref"]
        })
    }
    async fn execute(&self, args: Value) -> Result<String> {
        let pid = urlencoding::encode(
            args["project_id"]
                .as_str()
                .ok_or_else(|| anyhow::anyhow!("project_id required"))?,
        );
        let v: Value = self
            .client
            .post(
                &format!("projects/{}/pipeline", pid),
                json!({ "ref": args["ref"] }),
            )
            .await?;
        Ok(serde_json::to_string_pretty(&v)?)
    }
}

struct RetryPipeline {
    client: GitLabClient,
}
#[async_trait]
impl Tool for RetryPipeline {
    fn name(&self) -> &str {
        "retry_pipeline"
    }
    fn description(&self) -> &str {
        "Retry a failed pipeline."
    }
    fn parameters_schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "project_id": { "type": "string" },
                "pipeline_id": { "type": "integer" }
            },
            "required": ["project_id", "pipeline_id"]
        })
    }
    async fn execute(&self, args: Value) -> Result<String> {
        let pid = urlencoding::encode(
            args["project_id"]
                .as_str()
                .ok_or_else(|| anyhow::anyhow!("project_id required"))?,
        );
        let pipeline_id = args["pipeline_id"]
            .as_u64()
            .ok_or_else(|| anyhow::anyhow!("pipeline_id required"))?;
        let v: Value = self
            .client
            .post(
                &format!("projects/{}/pipelines/{}/retry", pid, pipeline_id),
                json!({}),
            )
            .await?;
        Ok(serde_json::to_string_pretty(&v)?)
    }
}

struct CancelPipeline {
    client: GitLabClient,
}
#[async_trait]
impl Tool for CancelPipeline {
    fn name(&self) -> &str {
        "cancel_pipeline"
    }
    fn description(&self) -> &str {
        "Cancel a running pipeline."
    }
    fn parameters_schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "project_id": { "type": "string" },
                "pipeline_id": { "type": "integer" }
            },
            "required": ["project_id", "pipeline_id"]
        })
    }
    async fn execute(&self, args: Value) -> Result<String> {
        let pid = urlencoding::encode(
            args["project_id"]
                .as_str()
                .ok_or_else(|| anyhow::anyhow!("project_id required"))?,
        );
        let pipeline_id = args["pipeline_id"]
            .as_u64()
            .ok_or_else(|| anyhow::anyhow!("pipeline_id required"))?;
        let v: Value = self
            .client
            .post(
                &format!("projects/{}/pipelines/{}/cancel", pid, pipeline_id),
                json!({}),
            )
            .await?;
        Ok(serde_json::to_string_pretty(&v)?)
    }
}

struct GetPipelineJobs {
    client: GitLabClient,
}
#[async_trait]
impl Tool for GetPipelineJobs {
    fn name(&self) -> &str {
        "get_pipeline_jobs"
    }
    fn description(&self) -> &str {
        "List all jobs for a specific pipeline."
    }
    fn parameters_schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "project_id": { "type": "string" },
                "pipeline_id": { "type": "integer" },
                "per_page": { "type": "integer", "default": 100 }
            },
            "required": ["project_id", "pipeline_id"]
        })
    }
    async fn execute(&self, args: Value) -> Result<String> {
        let pid = urlencoding::encode(
            args["project_id"]
                .as_str()
                .ok_or_else(|| anyhow::anyhow!("project_id required"))?,
        );
        let pipeline_id = args["pipeline_id"]
            .as_u64()
            .ok_or_else(|| anyhow::anyhow!("pipeline_id required"))?;
        let per_page = args["per_page"].as_u64().unwrap_or(100);
        let v: Vec<Value> = self
            .client
            .get(&format!(
                "projects/{}/pipelines/{}/jobs?per_page={}",
                pid, pipeline_id, per_page
            ))
            .await?;
        Ok(serde_json::to_string_pretty(&v)?)
    }
}

struct GetJobLog {
    client: GitLabClient,
}
#[async_trait]
impl Tool for GetJobLog {
    fn name(&self) -> &str {
        "get_job_log"
    }
    fn description(&self) -> &str {
        "Get the log/trace of a CI job."
    }
    fn parameters_schema(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "project_id": { "type": "string" },
                "job_id": { "type": "integer" }
            },
            "required": ["project_id", "job_id"]
        })
    }
    async fn execute(&self, args: Value) -> Result<String> {
        let pid = urlencoding::encode(
            args["project_id"]
                .as_str()
                .ok_or_else(|| anyhow::anyhow!("project_id required"))?,
        );
        let job_id = args["job_id"]
            .as_u64()
            .ok_or_else(|| anyhow::anyhow!("job_id required"))?;
        let resp = self
            .client
            .get_raw(
                &self
                    .client
                    .api_url(&format!("projects/{}/jobs/{}/trace", pid, job_id)),
            )
            .await?;
        Ok(resp.text().await?)
    }
}

File: extension/server/tools/pipelines.ts
import type { GitLabClient } from '../gitlabClient';
import type { Tool } from './tool';

function enc(s: string): string {
  return encodeURIComponent(s);
}

const ANSI_RE = /\x1b\[[0-9;]*[mGKHF]/g;
const ERROR_RE = /ERROR|FAILED|[Ee]rror:|fatal:|FATAL|AssertionError|Traceback|npm ERR!|exit code [1-9]/;

export function extractLogErrors(rawLog: string, maxLines: number, contextLines: number): string {
  const clean = rawLog.replace(ANSI_RE, "");
  const lines = clean.split("\n");
  const total = lines.length;

  // Find all lines matching error patterns
  const errorIndexes: number[] = [];
  for (let i = 0; i < lines.length; i++) {
    if (ERROR_RE.test(lines[i])) errorIndexes.push(i);
  }

  let selectedLines: string[];

  if (errorIndexes.length === 0) {
    // Fallback: last max_lines lines
    selectedLines = lines.slice(Math.max(0, total - maxLines));
  } else {
    // Build merged context windows
    const included = new Set<number>();
    for (const idx of errorIndexes) {
      const start = Math.max(0, idx - contextLines);
      const end = Math.min(total - 1, idx + contextLines);
      for (let i = start; i <= end; i++) included.add(i);
    }
    const sortedIndexes = Array.from(included).sort((a, b) => a - b);
    selectedLines = sortedIndexes.map(i => lines[i]);
  }

  // Keep the trailing window — the end of a failed log is most diagnostic
  if (selectedLines.length > maxLines) {
    selectedLines = selectedLines.slice(selectedLines.length - maxLines);
  }

  const header = `[Extracted ${errorIndexes.length} error lines from ${total} total.]`;
  return [header, ...selectedLines].join("\n");
}

export function pipelineTools(client: GitLabClient): Tool[] {
  return [
    {
      name: 'list_pipelines',
      description: 'List pipelines for a project.',
      parametersSchema: () => ({
        type: 'object',
        properties: {
          project_id: { type: 'string' },
          per_page: { type: 'integer', default: 20 },
        },
        required: ['project_id'],
      }),
      async execute(args) {
        const pid = enc(args.project_id as string);
        const perPage = (args.per_page as number) ?? 20;
        return JSON.stringify(await client.get(`projects/${pid}/pipelines?per_page=${perPage}`), null, 2);
      },
    },
    {
      name: 'get_pipeline',
      description: 'Get details of a specific pipeline.',
      parametersSchema: () => ({
        type: 'object',
        properties: {
          project_id: { type: 'string' },
          pipeline_id: { type: 'integer' },
        },
        required: ['project_id', 'pipeline_id'],
      }),
      async execute(args) {
        const pid = enc(args.project_id as string);
        return JSON.stringify(await client.get(`projects/${pid}/pipelines/${args.pipeline_id}`), null, 2);
      },
    },
    {
      name: 'trigger_pipeline',
      description: 'Trigger a new pipeline for a ref.',
      parametersSchema: () => ({
        type: 'object',
        properties: {
          project_id: { type: 'string' },
          ref: { type: 'string', description: 'Branch or tag name' },
        },
        required: ['project_id', 'ref'],
      }),
      async execute(args) {
        const pid = enc(args.project_id as string);
        return JSON.stringify(await client.post(`projects/${pid}/pipeline`, { ref: args.ref }), null, 2);
      },
    },
    {
      name: 'retry_pipeline',
      description: 'Retry a failed pipeline.',
      parametersSchema: () => ({
        type: 'object',
        properties: {
          project_id: { type: 'string' },
          pipeline_id: { type: 'integer' },
        },
        required: ['project_id', 'pipeline_id'],
      }),
      async execute(args) {
        const pid = enc(args.project_id as string);
        return JSON.stringify(
          await client.post(`projects/${pid}/pipelines/${args.pipeline_id}/retry`, {}),
          null, 2
        );
      },
    },
    {
      name: 'cancel_pipeline',
      description: 'Cancel a running pipeline.',
      parametersSchema: () => ({
        type: 'object',
        properties: {
          project_id: { type: 'string' },
          pipeline_id: { type: 'integer' },
        },
        required: ['project_id', 'pipeline_id'],
      }),
      async execute(args) {
        const pid = enc(args.project_id as string);
        return JSON.stringify(
          await client.post(`projects/${pid}/pipelines/${args.pipeline_id}/cancel`, {}),
          null, 2
        );
      },
    },
    {
      name: 'get_pipeline_jobs',
      description: 'List all jobs for a specific pipeline.',
      parametersSchema: () => ({
        type: 'object',
        properties: {
          project_id: { type: 'string' },
          pipeline_id: { type: 'integer' },
          per_page: { type: 'integer', default: 100 },
        },
        required: ['project_id', 'pipeline_id'],
      }),
      async execute(args) {
        const pid = enc(args.project_id as string);
        const perPage = (args.per_page as number) ?? 100;
        return JSON.stringify(
          await client.get(`projects/${pid}/pipelines/${args.pipeline_id}/jobs?per_page=${perPage}`),
          null, 2
        );
      },
    },
    {
      name: 'get_job_log',
      description: 'Get the log/trace of a CI job.',
      parametersSchema: () => ({
        type: 'object',
        properties: {
          project_id: { type: 'string' },
          job_id: { type: 'integer' },
        },
        required: ['project_id', 'job_id'],
      }),
      async execute(args) {
        const pid = enc(args.project_id as string);
        return client.getRaw(client.apiUrl(`projects/${pid}/jobs/${args.job_id}/trace`));
      },
    },
    {
      name: 'get_job_log_errors',
      description:
        'Get the relevant error/failure section of a CI job log. Strips ANSI codes, ' +
        'finds lines matching error patterns, returns them with context. Much smaller ' +
        'than get_job_log — use this for pipeline debugging.',
      parametersSchema: () => ({
        type: 'object',
        properties: {
          project_id: { type: 'string' },
          job_id: { type: 'integer' },
          max_lines: { type: 'integer', default: 150, description: 'Max lines to return' },
          context_lines: { type: 'integer', default: 5, description: 'Lines of context around each error' },
        },
        required: ['project_id', 'job_id'],
      }),
      async execute(args) {
        const pid = enc(args.project_id as string);
        const maxLines = (args.max_lines as number) ?? 150;
        const contextLines = (args.context_lines as number) ?? 5;
        const raw = await client.getRaw(client.apiUrl(`projects/${pid}/jobs/${args.job_id}/trace`));
        return extractLogErrors(raw, maxLines, contextLines);
      },
    },
  ];
}
